server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -860,8 +876,13 @@ public void onFailure(Exception e) {
                         listener.onFailure(e);
                     }
                 }
-            }, 30L, TimeUnit.MINUTES, EsExecutors.DIRECT_EXECUTOR_SERVICE); // Wait on current thread because this execution is wrapped by
-                                                                            // CancellableThreads and we want to be able to interrupt it
+            };
+            if (acquiredPrimaryPermits == null) {
+                // Wait on current thread because this execution is wrapped by CancellableThreads and we want to be able to interrupt it
+                indexShardOperationPermits.blockOperations(onAcquired, 30L, TimeUnit.MINUTES, EsExecutors.DIRECT_EXECUTOR_SERVICE);
+            } else {
+                ActionListener.completeWith(onAcquired, () -> acquiredPrimaryPermits);
```

Comment:
can we at least assert that all the primary permits are acquired if `acquiredPrimaryPermits != null`?

Reply:
Done!